text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,705 @@
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.0
+    }
+}
+
+trait MyTrait {}
+
+impl MyTrait for i32 {}
+
+fn main() {
+    let ptr: MySmartPointer<i32> = MySmartPointer(Box::new(4));
+
+    // This coercion would be an error without the derive.
+    let ptr: MySmartPointer<dyn MyTrait> = ptr;
+}
+```
+Without the `#[derive(SmartPointer)]` macro, this example would fail with the
+following error:
+```
+error[E0308]: mismatched types
+  --> src/main.rs:11:44
+   |
+11 |     let ptr: MySmartPointer<dyn MyTrait> = ptr;
+   |              ---------------------------   ^^^ expected `MySmartPointer<dyn MyTrait>`, found `MySmartPointer<i32>`
+   |              |
+   |              expected due to this
+   |
+   = note: expected struct `MySmartPointer<dyn MyTrait>`
+              found struct `MySmartPointer<i32>`
+   = help: `i32` implements `MyTrait` so you could box the found value and coerce it to the trait object `Box<dyn MyTrait>`, you will have to change the expected type as well
+```
+
+## Object safety
+
+Consider the following trait:
+```rust
+trait MyTrait {
+    // Arbitrary self types is enough for this.
+    fn func(self: MySmartPointer<Self>);
+}
+
+// But this requires #[derive(SmartPointer)].
+fn call_func(value: MySmartPointer<dyn MyTrait>) {
+    value.func();
+}
+```
+You do not need `#[derive(SmartPointer)]` to declare this trait ([arbitrary
+self types][ast] is enough), but the trait will not be object safe unless you
+annotate `MySmartPointer` with `#[derive(SmartPointer)]`. If you don't, then
+the use of `dyn MyTrait` triggers the following error:
+```
+error[E0038]: the trait `MyTrait` cannot be made into an object
+  --> src/lib.rs:11:36
+   |
+8  |     fn func(self: MySmartPointer<Self>);
+   |                   -------------------- help: consider changing method `func`'s `self` parameter to be `&self`: `&Self`
+...
+11 | fn call_func(value: MySmartPointer<dyn MyTrait>) {
+   |                                    ^^^^^^^^^^^ `MyTrait` cannot be made into an object
+   |
+note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
+  --> src/lib.rs:8:19
+   |
+7  | trait MyTrait {
+   |       ------- this trait cannot be made into an object...
+8  |     fn func(self: MySmartPointer<Self>);
+   |                   ^^^^^^^^^^^^^^^^^^^^ ...because method `func`'s `self` parameter cannot be dispatched on
+```
+Note that using the `self: MySmartPointer<Self>` syntax requires that you
+implement `Receiver` (or `Deref`), as the derive macro does not emit an
+implementation of `Receiver`.
+
+## Requirements for using the macro
+
+Whenever a `self: MySmartPointer<Self>` method is called on a trait object, the
+compiler will convert from `MySmartPointer<dyn MyTrait>` to
+`MySmartPointer<MyStruct>` using something similar to a transmute. Because of
```

Comment:
It's actually worse than a transmute of these values, we're effectively transmuting the function pointer. This means the two types do not just have to be layout compatible, they have to be [function call ABI compatible](https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility). That's also why allowing extra fields in these types is pretty difficult.